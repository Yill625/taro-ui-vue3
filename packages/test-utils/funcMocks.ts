import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks