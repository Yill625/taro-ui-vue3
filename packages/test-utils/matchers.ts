import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers