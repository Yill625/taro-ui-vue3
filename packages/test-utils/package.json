{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.2.0",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}