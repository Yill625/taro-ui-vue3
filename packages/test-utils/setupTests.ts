import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()