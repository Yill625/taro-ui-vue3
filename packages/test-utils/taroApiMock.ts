import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
