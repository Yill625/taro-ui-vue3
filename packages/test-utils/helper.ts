import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}