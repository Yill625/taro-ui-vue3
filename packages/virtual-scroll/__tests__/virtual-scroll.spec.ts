import { h } from '@vue/runtime-core'
import { mountFactory, Slots } from '@taro-ui-vue3/test-utils/helper'
import AtVirtualScroll from '../index'
import Taro from '@tarojs/taro'

const elementHeight = 100

const mountFn = (
  props = {},
  slots: Slots = {
    default: ({ index, item }) => {
      return [h('view', { class: 'item' }, `${index}: item-${item}`)]
    }
  }
) => {
  return mountFactory(AtVirtualScroll, undefined, props, slots)
}

const props = {
  height: elementHeight,
  items: [1, 2, 3],
  itemHeight: 50,
}

describe('AtVirtualScroll', () => {
  let mock: jest.SpyInstance

  beforeEach(() => {
    // mock Taro.getEnv, and return a non-WEB environment
    jest.mock('@tarojs/taro')
    jest.spyOn(Taro, 'getEnv').mockImplementation(() => {
      return Taro.ENV_TYPE.WEAPP
    })

    // mock clientHeight
    mock = jest.spyOn(window.HTMLElement.prototype, 'clientHeight', 'get').mockReturnValue(elementHeight)
  })

  afterEach(() => {
    mock.mockRestore()
  })

  it('should render component with scopedSlot and match snapshot', async () => {
    const wrapper = mountFn(props)

    await wrapper.vm.$nextTick()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should set height of scrollable element', () => {
    const wrapper = mountFn(props)

    const scrollable = wrapper.find('.at-virtual-scroll__container')
    expect(scrollable.attributes('style')).toEqual('height: 150px;')
  })

  it('should render not more than 5 hidden items and match snapshot', async () => {
    const wrapper = mountFn({
      height: elementHeight,
      items: [1, 2, 3, 4, 5, 6, 7, 8, 9, 0],
      itemHeight: 50
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render right items on scroll and match snapshot', async () => {
    const wrapper = mountFn({
      height: elementHeight,
      items: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
      itemHeight: 50
    })

    await wrapper.vm.$nextTick()
    await wrapper.find('.at-virtual-scroll').trigger('scroll', { detail: { scrollTop: 500 } })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should scroll to the provided item index', async () => {
    const wrapper = mountFn({
      height: elementHeight,
      items: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
      itemHeight: 50,
      scrollIntoItem: 10
    })

    await wrapper.vm.$nextTick()

    expect(wrapper.find('.item').text()).toBe('10: item-11')
    expect(wrapper.element).toMatchSnapshot()

    await wrapper.setProps({ scrollIntoItem: 15 })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.item').text()).toBe('15: item-16')
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render header slot', async () => {
    const wrapper = mountFn(props, {
      default: ({ index, item }) => [h('view', { class: 'item' }, `${index}: item-${item}`)],
      header: () => [h('view')]
    })

    await wrapper.vm.$nextTick()
    expect(wrapper.find('.at-virtual-scroll__header > view').exists()).toBe(true)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render footer slot', async () => {
    const wrapper = mountFn(props, {
      default: ({ index, item }) => [h('view', { class: 'item' }, `${index}: item-${item}`)],
      footer: () => [h('view')]
    })

    await wrapper.vm.$nextTick()
    expect(wrapper.find('.at-virtual-scroll__footer > view').exists()).toBe(true)
    expect(wrapper.element).toMatchSnapshot()
  })
})

describe('AtVirtualScroll Behavior', () => {
  it('should trigger onReachTop event', async () => {
    const onReachTop = jest.fn()

    const wrapper = mountFn({
      ...props,
      onReachTop: onReachTop
    })

    await wrapper.vm.$nextTick()

    await wrapper.find('.at-virtual-scroll').trigger('scrolltoupper')
    expect(onReachTop).toBeCalled()
  })

  it('should trigger onReachBottom event', async () => {
    const onReachBottom = jest.fn()

    const wrapper = mountFn({
      ...props,
      onReachBottom: onReachBottom
    })

    await wrapper.vm.$nextTick()

    await wrapper.find('.at-virtual-scroll').trigger('scrolltolower')
    expect(onReachBottom).toBeCalled()
  })
})