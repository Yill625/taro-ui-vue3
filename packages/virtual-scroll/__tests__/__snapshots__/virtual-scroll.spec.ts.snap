// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtVirtualScroll should render component with scopedSlot and match snapshot 1`] = `
<view>
  <!---->
  <scroll-view
    class="at-virtual-scroll"
    lowerthreshold="50"
    scrollintoview="item-undefined"
    scrollwithanimation="true"
    scrolly="true"
    style="height: 100px;"
    upperthreshold="50"
  >
    <view>
      <view
        class="at-virtual-scroll__container"
        style="height: 150px;"
      >
        <view
          class="at-virtual-scroll__item"
          id="item-0"
          style="top: 0px;"
        >
          <view
            class="item"
          >
            0: item-1
          </view>
        </view>
        <view
          class="at-virtual-scroll__item"
          id="item-1"
          style="top: 50px;"
        >
          <view
            class="item"
          >
            1: item-2
          </view>
        </view>
      </view>
      <!---->
    </view>
  </scroll-view>
</view>
`;

exports[`AtVirtualScroll should render footer slot 1`] = `
<view>
  <!---->
  <scroll-view
    class="at-virtual-scroll"
    lowerthreshold="50"
    scrollintoview="item-undefined"
    scrollwithanimation="true"
    scrolly="true"
    style="height: 100px;"
    upperthreshold="50"
  >
    <view>
      <view
        class="at-virtual-scroll__container"
        style="height: 150px;"
      >
        <view
          class="at-virtual-scroll__item"
          id="item-0"
          style="top: 0px;"
        >
          <view
            class="item"
          >
            0: item-1
          </view>
        </view>
        <view
          class="at-virtual-scroll__item"
          id="item-1"
          style="top: 50px;"
        >
          <view
            class="item"
          >
            1: item-2
          </view>
        </view>
      </view>
      <view
        class="at-virtual-scroll__footer"
      >
        <view />
      </view>
    </view>
  </scroll-view>
</view>
`;

exports[`AtVirtualScroll should render header slot 1`] = `
<view>
  <view
    class="at-virtual-scroll__header"
  >
    <view />
  </view>
  <scroll-view
    class="at-virtual-scroll"
    lowerthreshold="50"
    scrollintoview="item-undefined"
    scrollwithanimation="true"
    scrolly="true"
    style="height: 100px;"
    upperthreshold="50"
  >
    <view>
      <view
        class="at-virtual-scroll__container"
        style="height: 150px;"
      >
        <view
          class="at-virtual-scroll__item"
          id="item-0"
          style="top: 0px;"
        >
          <view
            class="item"
          >
            0: item-1
          </view>
        </view>
        <view
          class="at-virtual-scroll__item"
          id="item-1"
          style="top: 50px;"
        >
          <view
            class="item"
          >
            1: item-2
          </view>
        </view>
      </view>
      <!---->
    </view>
  </scroll-view>
</view>
`;

exports[`AtVirtualScroll should render not more than 5 hidden items and match snapshot 1`] = `
<view>
  <!---->
  <scroll-view
    class="at-virtual-scroll"
    lowerthreshold="50"
    scrollintoview="item-undefined"
    scrollwithanimation="true"
    scrolly="true"
    style="height: 100px;"
    upperthreshold="50"
  >
    <view>
      <view
        class="at-virtual-scroll__container"
        style="height: 500px;"
      >
        <view
          class="at-virtual-scroll__item"
          id="item-0"
          style="top: 0px;"
        >
          <view
            class="item"
          >
            0: item-1
          </view>
        </view>
        <view
          class="at-virtual-scroll__item"
          id="item-1"
          style="top: 50px;"
        >
          <view
            class="item"
          >
            1: item-2
          </view>
        </view>
      </view>
      <!---->
    </view>
  </scroll-view>
</view>
`;

exports[`AtVirtualScroll should render right items on scroll and match snapshot 1`] = `
<view>
  <!---->
  <scroll-view
    class="at-virtual-scroll"
    lowerthreshold="50"
    scrollintoview="item-undefined"
    scrollwithanimation="true"
    scrolly="true"
    style="height: 100px;"
    upperthreshold="50"
  >
    <view>
      <view
        class="at-virtual-scroll__container"
        style="height: 1000px;"
      />
      <!---->
    </view>
  </scroll-view>
</view>
`;

exports[`AtVirtualScroll should scroll to the provided item index 1`] = `
<view>
  <!---->
  <scroll-view
    class="at-virtual-scroll"
    lowerthreshold="50"
    scrollintoview="item-10"
    scrollwithanimation="true"
    scrolly="true"
    style="height: 100px;"
    upperthreshold="50"
  >
    <view>
      <view
        class="at-virtual-scroll__container"
        style="height: 1000px;"
      >
        <view
          class="at-virtual-scroll__item"
          id="item-10"
          style="top: 500px;"
        >
          <view
            class="item"
          >
            10: item-11
          </view>
        </view>
        <view
          class="at-virtual-scroll__item"
          id="item-11"
          style="top: 550px;"
        >
          <view
            class="item"
          >
            11: item-12
          </view>
        </view>
      </view>
      <!---->
    </view>
  </scroll-view>
</view>
`;

exports[`AtVirtualScroll should scroll to the provided item index 2`] = `
<view>
  <!---->
  <scroll-view
    class="at-virtual-scroll"
    lowerthreshold="50"
    scrollintoview="item-15"
    scrollwithanimation="true"
    scrolly="true"
    style="height: 100px;"
    upperthreshold="50"
  >
    <view>
      <view
        class="at-virtual-scroll__container"
        style="height: 1000px;"
      >
        <view
          class="at-virtual-scroll__item"
          id="item-15"
          style="top: 750px;"
        >
          <view
            class="item"
          >
            15: item-16
          </view>
        </view>
        <view
          class="at-virtual-scroll__item"
          id="item-16"
          style="top: 800px;"
        >
          <view
            class="item"
          >
            16: item-17
          </view>
        </view>
      </view>
      <!---->
    </view>
  </scroll-view>
</view>
`;
